File: 1160-find-words-that-can-be-formed-by-characters/1160-find-words-that-can-be-formed-by-characters.go
func countCharacters(words []string, chars string) int {

    charsMap := make(map[string]int)
    result := 0

    for i := 0; i < len(chars); i++ {
        charsMap[string(chars[i])]++
    }

    for j := 0; j < len(words); j++ {
        charsMapCopy := make(map[string]int)
        for k, v := range charsMap {
            charsMapCopy[k] = v
        }
        valid := true

        for k := 0; k < len(words[j]); k++ {
            _, exists := charsMapCopy[string(words[j][k])]
            if ((!exists) || (charsMapCopy[string(words[j][k])] - 1 < 0)) {
                valid = false
                break
            }
            charsMapCopy[string(words[j][k])]--
        }

        if valid {
            result += len(words[j])
        }
    }

    return result
}

File: 1160-find-words-that-can-be-formed-by-characters/1160-find-words-that-can-be-formed-by-characters.cpp
class Solution {
public:
    int countCharacters(vector<string>& words, string chars) {
        map<char, int> charCounts;
        for (const char c : chars) {
            ++charCounts[c];
        }
        int result = 0;
        for (const string word : words) {
            map<char, int> wordCharCounts;
            bool canFormWord = true;
            for (const char c : word) {
                if (!charCounts.contains(c)) {
                    canFormWord = false;
                    break;
                }
                ++wordCharCounts[c];
                if (wordCharCounts[c] > charCounts[c]) {
                    canFormWord = false;
                    break;
                }
            }
            if (canFormWord) result += word.size();
        }
        return result;
    }
};